ext/json/json_parser.y

```diff
@@ -255,10 +255,40 @@ static int php_json_parser_object_update(php_json_parser *parser, zval *object,
 	return SUCCESS;
 }
 
+static int php_json_parser_array_create_validate(php_json_parser *parser, zval *array)
+{
+	ZVAL_EMPTY_ARRAY(array);
+	return SUCCESS;
+}
+
+static int php_json_parser_array_append_validate(php_json_parser *parser, zval *array, zval *zvalue)
+{
+	return SUCCESS;
+}
+
+static int php_json_parser_object_create_validate(php_json_parser *parser, zval *object)
+{
+	ZVAL_EMPTY_ARRAY(object);
+	return SUCCESS;
+}
+
+static int php_json_parser_object_update_validate(php_json_parser *parser, zval *object, zend_string *key, zval *zvalue)
+{
+	return SUCCESS;
+}
+
 static int php_json_yylex(union YYSTYPE *value, php_json_parser *parser)
 {
 	int token = php_json_scan(&parser->scanner);
 	value->value = parser->scanner.value;
+
+	if (parser->methods.array_create == php_json_parser_array_create_validate
+		&& parser->methods.array_append == php_json_parser_array_append_validate
+		&& parser->methods.object_create == php_json_parser_object_create_validate
+		&& parser->methods.object_update == php_json_parser_object_update_validate) {
+		zval_ptr_dtor_str(&(parser->scanner.value));
```

Comment:
nit: can optimize slightly
```suggestion
    // php_json_parser_array_create_validate is a C static function, the only caller using it is php_json_validate_ex
	if (parser->methods.array_create == php_json_parser_array_create_validate) {
		zval_ptr_dtor_str(&(parser->scanner.value));
```

Reply:
The observation is right, but I want to keep it consistent for future changes too. I don't know how this code will change in the future.

The way I wrote it makes sure that "validation only" is being done.

I also thought about creating a new variable indicating the "execution context" of the parser, so we would be able to know:

- the parser is being use by json_decode()
- the parser is being use by json_validate()
- etc.

and use such a "execution context" in this IF block like ...

```C
if (execution_context === "json_validate") {
    zval_ptr_dtor_str(&(parser->scanner.value));
}
```
But then I gave back and I thought might not like to most of developers.

Honestly I dont know if I should apply your suggestion or not, I am in open position here.

Maybe someone else can give an opinion about this.

Reply:
This is called just once so probably not a huge optimization IMO so I think it's fine to leave it as it is.

Reply:
> This is called just once so probably not a huge optimization IMO so I think it's fine to leave it as it is.

It's called **once per json token** - the return value of this function and the return value of this and `int php_json_scan` is the token/error type (in the json_scanner.c for `PHP_JSON_*` constants having the token type)

- a json token is `[`, `{`, `123`, `false`, `"key"`, etc

Bison generates code that repeatedly calls this on the instantiated scanner while parsing 

So it'd have little effect on json_decode, but a larger effect on json_validate()